### PR TITLE
Native support for the Dot Pad tactile graphic Braille Display from Dot Inc

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None

### Summary of the issue:
The Dot Pad from Dot Inc is a device that can display refreshable tactile graphics and braille. The A320 model can show tactile graphics of 120 by 80 dots (which could comfortably fit 8 lines of 20 cells each. 
The device also has a dedicated line of 20 cells specifically for text.
Currently you can use the Dot Pad with NVDA via Brltty I believe, though I did not manage to get that working on my machine. There are also some NVDA add-ons to support Dot Pad in various forms,  using the Dot Pad SDK.

NVDA should add native support for the Dot pad device, removing the requirement for Brltty or add-ons, at least for basic functionality.

### Description of user facing changes
It is now possible to use the Dot Pad braille display with NVDA.
Features:
* Can display braille either on the dedicated 20 cell text line, or on the tactile graphic area. This can be configured in NVDA's braille settings panel by changing the Braille Destination combo box.
* Left and right panning keys are supported.

### Description of development approach
Write a native braille display driver that communicates with the device over serial. Borrowed many of the constants and structures from BrlTty.
Also added a new `tactile` package, which contains untilities and classes for working with tactile graphics. So far just a `TactileGraphicsBuffer` which represents  a buffer in which single dots can be set, and a `drawBrailleCells` function which can take bytes representing a string of standard braille cells, and draws them onto the graphics buffer.

### Testing strategy:
Tested with a Dot pad A320 via USB. Used NVDA normally with braille over a period of time.

### Known issues with pull request:
* No auto detection
* Dot pad will not refresh correctly if your hand is currently on the device. This is a hardware limitation.
* NvDA's current multiline braille support is still very limited, in that it treats the graphics area as one long wrapped line. However, as NVDA internally does know the number of rows and columns, future improvemts such as per-line word wrapping are coming.
 
 
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Added native support for the Dot Pad tactile graphics device, enhancing braille output.
	- Introduced functionality for rendering Braille cells on a tactile graphics buffer.
	- Created an abstract class for managing tactile graphics buffers to support future developments.

- **Documentation**
	- Updated user guide with a new section dedicated to the Dot Pad, detailing its capabilities and configuration options.

### CODE REVIEW

**Comment by bramd** [comment]
@michaelDCurran I would love to see better support for the Dot Pad in core. However, I've been working on an add-on for Dot to do this for a while now. We have lots of stuff that is in my opinion experimental and does not belong in core yet. However, if there is interest in adding tactile display support to core from NVAccess I would be glad to discuss which parts could be in core and how we can work together on that. Feel free to get in touch outside of Github as well for a more in depth discussion.

Unfortunately the addon code base is not open source right now, but will of course release it under GPL when we have a first public release.

**Comment by michaelDCurran** [comment]
@bramd  thanks for making contact. I've sent you an email about collaborating.

File: source/brailleDisplayDrivers/dotPad.py [hunk 1]
```diff
@@ -0,0 +1,475 @@
+# A part of NonVisual Desktop Access (NVDA)
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:1]: Is it possible to structure this like `source\brailleDisplayDrivers\albatross` or `source\brailleDisplayDrivers\eurobraille`?

File: source/brailleDisplayDrivers/dotPad.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+
+class DP_DisplayResponse(enum.IntEnum):
+	ACK = 0
+	NACK = 1
+	WAIT = 2
+	CHECKSUM = 3
+
+
+class DP_Features(enum.IntEnum):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:63]:
```suggestion
class DP_Features(enum.Flag):
```

or

```suggestion
class DP_Features(enum.IntFlag):
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+	def __init__(self, hCellCount: int, vCellCount: int):
+		self.hCellCount = hCellCount
+		self.vCellCount = vCellCount
+		self._cellBuffer = bytearray(hCellCount * vCellCount)
+		hPixelCount = hCellCount * self.cellWidth
+		vPixelCount = vCellCount * self.cellHeight
+		super().__init__(hPixelCount, vPixelCount)
+
+	def setDot(self, x: int, y: int):
+		if x < 0 or x >= self.width or y < 0 or y >= self.height:
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:141]:
```suggestion
		if not (0 <= x < self.width) or not (0 <= y < self.height):
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+class CommandResponse:
+	cmd: DP_Command
+	data: bytes
+	dest: int
+	seqNum: int
+
+
+class BrailleDestination(str, enum.Enum):
+	text = "text"
+	graphic = "graphic"
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:164]: Constants should be upper case

```suggestion
class BrailleDestination(enum.StrEnum):
	TEXT = "text"
	GRAPHIC = "graphic"
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+
+
+class BrailleDisplayDriver(braille.BrailleDisplayDriver):
+	"""
+	Driver for DotPad Braille / Tactile Graphic display.
+	"""
+
+	name = "dotPad"
+	# Translators: Names of braille displays.
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:173]: this comment seems incorrectly placed

File: source/brailleDisplayDrivers/dotPad.py [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+	_lastResponse: dict[int, CommandResponse] = {}
+
+	def _sendCommand(
+		self,
+		cmd: DP_Command,
+		data: bytes = b"",
+		dest: int = 0,
+		seqNum: int = 0,
+		rspCmd: DP_Command | None = None,
+	):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:209]:
```suggestion
	) -> bytes:
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+				if response is not None and response.cmd == rspCmd and response.dest == dest:
+					break
+				if x > 0:
+					ctypes.windll.kernel32.SleepEx(50, True)
+			else:
+				raise RuntimeError(f"No response to {cmd.name}")
+			return response.data
+		return b""
+
+	def _onReceive(self, header1):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:243]:
```suggestion
	def _onReceive(self, header1: bytes):
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+		log.debug(f"Received responce  {cmd.name}, {dest=}, {seqNum=}, data={bytes(data)}")
+		if cmd.name.startswith("RSP_"):
+			self._recordCommandResponse(cmd, data, dest, seqNum)
+		elif cmd.name.startswith("NTF_"):
+			self._handleNotification(cmd, data, dest, seqNum)
+
+	def _recordCommandResponse(self, cmd: DP_Command, data: bytes, dest: int = 0, seqNum: int = 0):
+		self._lastResponse[dest] = CommandResponse(cmd, data, dest, seqNum)
+
+	def _requestDeviceName(self):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:265]:
```suggestion
	def _requestDeviceName(self) -> str:
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+		data = self._sendCommand(DP_Command.REQ_DEVICE_NAME, rspCmd=DP_Command.RSP_DEVICE_NAME)
+		return data.decode("ascii")
+
+	def _requestBoardInformation(self):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:269]:
```suggestion
	def _requestBoardInformation(self) -> DP_BoardInformation:
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+		data = self._sendCommand(DP_Command.REQ_BOARD_INFORMATION, rspCmd=DP_Command.RSP_BOARD_INFORMATION)
+		return DP_BoardInformation.from_buffer_copy(data)
+
+	_displayLineCache = {}
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:273]: can this get type information added?

File: source/brailleDisplayDrivers/dotPad.py [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+			self._brailleDestination = BrailleDestination.text
+		elif self._boardInformation.features & DP_Features.HAS_GRAPHIC_DISPLAY:
+			self._brailleDestination = BrailleDestination.graphic
+		else:
+			raise RuntimeError("No text or graphics displays")
+		super().__init__()
+
+	def terminate(self):
+		try:
+			super(BrailleDisplayDriver, self).terminate()
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:330]:
```suggestion
			super().terminate()
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+			return {}
+		dests = {}
+		if self._boardInformation.features & DP_Features.HAS_TEXT_DISPLAY:
+			dests["text"] = StringParameterInfo(
+				BrailleDestination.text.value,
+				# Translators: A destination for braille output.
+				_("Text"),
+			)
+		if self._boardInformation.features & DP_Features.HAS_GRAPHIC_DISPLAY:
+			dests["graphic"] = StringParameterInfo(
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:347]:
```suggestion
			dests[BrailleDestination.graphic.value] = StringParameterInfo(
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+			# Make sure the device gets closed.
+			# If it doesn't, we may not be able to re-open it later.
+			self._dev.close()
+
+	def _get_availableBrailledestinations(self):
+		if self._boardInformation is None:
+			return {}
+		dests = {}
+		if self._boardInformation.features & DP_Features.HAS_TEXT_DISPLAY:
+			dests["text"] = StringParameterInfo(
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:341]:
```suggestion
			dests[BrailleDestination.TEXT.value] = StringParameterInfo(
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+			)
+		if self._boardInformation.features & DP_Features.HAS_GRAPHIC_DISPLAY:
+			dests["graphic"] = StringParameterInfo(
+				BrailleDestination.graphic.value,
+				# Translators: A destination for braille output.
+				_("Graphic"),
+			)
+		return dests
+
+	def _get_brailleDestination(self):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:354]:
```suggestion
	def _get_brailleDestination(self) -> str:
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+			self._brailleDestination = value
+		elif (
+			value == BrailleDestination.graphic
+			and self._boardInformation.features & DP_Features.HAS_GRAPHIC_DISPLAY
+		):
+			self._brailleDestination = value
+		else:
+			raise ValueError(f"Unsupported destination {value}")
+
+	def _get_numRows(self):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:376]:
```suggestion
	def _get_numRows(self) -> int:
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+		if self._boardInformation is None:
+			return 0
+		if self._brailleDestination == BrailleDestination.graphic:
+			return (self._boardInformation.graphic.rowCount * self.cellHeight) // (
+				self.cellHeight + self.vCellPadding
+			)
+		else:
+			return self._boardInformation.text.rowCount
+
+	def _get_numCols(self):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:386]:
```suggestion
	def _get_numCols(self) -> int:
```

File: source/brailleDisplayDrivers/dotPad.py [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+			return 0
+		if self._brailleDestination == BrailleDestination.graphic:
+			return (self._boardInformation.graphic.columnCount * self.cellWidth) // (
+				self.cellWidth + self.hCellPadding
+			)
+		else:
+			return self._boardInformation.text.columnCount
+
+	_inDisplay = False
+	_pendingCells = None
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/dotPad.py:397]: can this get type info added?

File: source/tactile/braille.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+	:param y: The y coordinate of the top-left corner of the first cell.
+	:param cells: A list of 8-bit integers representing the Braille cells to draw.
+	:param hCellPadding: The horizontal padding between cells.
+	"""
+	for cell in cells:
+		for dot in range(0, 8):
+			if 1 << dot & cell:
+				dotX, dotY = _brailleDotCoords[dot]
+				tgBuf.setDot(x + dotX, y + dotY)
+		x += CELL_WIDTH + hCellPadding
```
[the remark this answers is not in the document]
Comment (michaelDCurran) [reply source/tactile/braille.py:51]: @coderabbitai please provide this as a Github code suggestion that can be applied.

File: user_docs/en/userGuide.md [hunk 19]
```diff
@@ -5059,6 +5059,31 @@ Following are the current key assignments for these displays.
 
 <!-- KC:endInclude -->
 
+### Dot Pad  {#dotPad}
```
Comment (seanbudd) [inline user_docs/en/userGuide.md:5062]:
```suggestion
### Dot Pad {#dotPad}
```

File: source/tactile/braille.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+	:param y: The y coordinate of the top-left corner of the first cell.
+	:param cells: A list of 8-bit integers representing the Braille cells to draw.
+	:param hCellPadding: The horizontal padding between cells.
+	"""
+	for cell in cells:
+		for dot in range(0, 8):
+			if 1 << dot & cell:
+				dotX, dotY = _brailleDotCoords[dot]
+				# Check if the coordinates are within bounds
+				if 0 <= x + dotX < tgBuf.width and 0 <= y + dotY < tgBuf.height:
```
Comment (seanbudd) [inline source/tactile/braille.py:51]: Just to improve readability

```suggestion
				if 0 <= (x + dotX) < tgBuf.width and 0 <= (y + dotY) < tgBuf.height:
```

```suggestion
				if (
					0 <= (x + dotX) < tgBuf.width
					and 0 <= (y + dotY) < tgBuf.height
				):
```

**Review by seanbudd — APPROVED** [review]

**Review by seanbudd — APPROVED** [review]

**Review by Qchristensen — APPROVED** [review]
Great work, will be a good asset to Dot Pad users :)

**Comment by SaschaCowley** [comment]
Can you please resolve the merge conflicts?

**Pull request merged**